File: sample-videochat-conference-java/src/main/java/com/quickblox/sample/videochat/conference/java/activities/DialogsActivity.java
package com.quickblox.sample.videochat.conference.java.activities;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.quickblox.chat.model.QBChatDialog;
import com.quickblox.conference.ConferenceClient;
import com.quickblox.conference.ConferenceSession;
import com.quickblox.conference.WsException;
import com.quickblox.conference.callbacks.ConferenceEntityCallback;
import com.quickblox.core.QBEntityCallback;
import com.quickblox.core.exception.QBResponseException;
import com.quickblox.sample.videochat.conference.java.R;
import com.quickblox.sample.videochat.conference.java.adapters.DialogsAdapter;
import com.quickblox.sample.videochat.conference.java.db.QbUsersDbManager;
import com.quickblox.sample.videochat.conference.java.fragments.ProgressDialogFragment;
import com.quickblox.sample.videochat.conference.java.utils.Consts;
import com.quickblox.sample.videochat.conference.java.utils.PermissionsChecker;
import com.quickblox.sample.videochat.conference.java.utils.SharedPrefsHelper;
import com.quickblox.sample.videochat.conference.java.utils.UsersUtils;
import com.quickblox.sample.videochat.conference.java.utils.WebRtcSessionManager;
import com.quickblox.users.model.QBUser;
import com.quickblox.videochat.webrtc.QBRTCTypes;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import androidx.appcompat.view.ActionMode;


public class DialogsActivity extends BaseActivity {
    private static final String TAG = DialogsActivity.class.getSimpleName();
    private static final int REQUEST_SELECT_PEOPLE = 174;
    private static final int REQUEST_PERMISSION = 175;

    private DialogsAdapter dialogsAdapter;
    private ListView dialogsListView;
    private QBUser currentUser;
    private ArrayList<QBChatDialog> chatDialogs;
    private QbUsersDbManager dbManager;
    private WebRtcSessionManager webRtcSessionManager;
    private ActionMode currentActionMode;
    private FloatingActionButton fab;
    private String dialogID;
    private List<Integer> occupants;
    private boolean isVideoCall;

    private PermissionsChecker checker;

    public static void start(Context context) {
        Intent intent = new Intent(context, DialogsActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_REORDER_TO_FRONT);
        context.startActivity(intent);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_dialogs);
        initFields();
        initDefaultActionBar();
        initUi();

        startLoadDialogs();
        loadUsersFromQb();

        checker = new PermissionsChecker(getApplicationContext());
    }

    private void initFields() {
        currentUser = sharedPrefsHelper.getQbUser();
        dbManager = QbUsersDbManager.getInstance(getApplicationContext());
        webRtcSessionManager = WebRtcSessionManager.getInstance(getApplicationContext());
    }

    private void initUi() {
        dialogsListView = findViewById(R.id.list_dialogs);
        fab = findViewById(R.id.fab_dialogs_new_chat);
    }

    private void startLoadDialogs() {
        showProgressDialog(R.string.dlg_loading_dialogs_users);
        requestExecutor.loadDialogs(new QBEntityCallback<ArrayList<QBChatDialog>>() {
            @Override
            public void onSuccess(ArrayList<QBChatDialog> result, Bundle params) {
                hideProgressDialog();
                chatDialogs = result;
                initDialogAdapter();
            }

            @Override
            public void onError(QBResponseException responseException) {
                hideProgressDialog();
                showErrorSnackbar(R.string.loading_users_error, responseException, new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        startLoadDialogs();
                    }
                });
            }
        });
    }

    private void loadUsersFromQb() {
        showProgressDialog(R.string.dlg_loading_dialogs_users);
        String currentRoomName = SharedPrefsHelper.getInstance().get(Consts.PREF_CURREN_ROOM_NAME);

        requestExecutor.loadUsersByTag(currentRoomName, new QBEntityCallback<ArrayList<QBUser>>() {
            @Override
            public void onSuccess(ArrayList<QBUser> result, Bundle params) {
                hideProgressDialog();
                dbManager.saveAllUsers(result, true);
            }

            @Override
            public void onError(QBResponseException responseException) {
                hideProgressDialog();
                showErrorSnackbar(R.string.loading_users_error, responseException, new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        loadUsersFromQb();
                    }
                });
            }
        });
    }

    @Override
    public ActionMode startSupportActionMode(ActionMode.Callback callback) {
        currentActionMode = super.startSupportActionMode(callback);
        return currentActionMode;
    }

    private void initDialogAdapter() {
        Log.d(TAG, "proceedInitUsersList chatDialogs= " + chatDialogs);
        if (dialogsAdapter == null) {
            dialogsAdapter = new DialogsAdapter(this, chatDialogs);
            dialogsListView.setAdapter(dialogsAdapter);
            dialogsListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                @Override
                public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                    QBChatDialog selectedDialog = (QBChatDialog) parent.getItemAtPosition(position);
                    if (currentActionMode == null) {
                        Log.d(TAG, "startConference selectedDialog.getDialogId()= " + selectedDialog.getDialogId()
                                + ", currentUser.getId()= " + currentUser.getId());
                        occupants = selectedDialog.getOccupants();
                        occupants.remove(currentUser.getId());
                        dialogID = selectedDialog.getDialogId();


                        dialogsAdapter.toggleOneItem(selectedDialog);
                        invalidateOptionsMenu();

                    } else {
                        dialogsAdapter.toggleSelection(selectedDialog);
                        updateActionBar(dialogsAdapter.getSelectedItems().size());
                    }
                }
            });
            dialogsListView.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
                @Override
                public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
                    QBChatDialog selectedDialog = (QBChatDialog) parent.getItemAtPosition(position);
                    startSupportActionMode(new DeleteActionModeCallback());
                    dialogsAdapter.selectItem(selectedDialog);
                    updateActionBar(dialogsAdapter.getSelectedItems().size());
                    return true;
                }
            });
        } else {
            dialogsAdapter.updateList(chatDialogs);
        }
    }

    private void startPermissionsActivity(boolean checkOnlyAudio) {
        PermissionsActivity.startForResult(this, REQUEST_PERMISSION, checkOnlyAudio, Consts.PERMISSIONS);
    }

    private void updateActionBar(int countSelectedUsers) {
        currentActionMode.setSubtitle(null);
        currentActionMode.setTitle(String.format(getString(
                countSelectedUsers > 1
                        ? R.string.tile_many_dialogs_selected
                        : R.string.title_one_dialog_selected),
                countSelectedUsers));
        currentActionMode.invalidate();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        if (dialogsAdapter != null && !dialogsAdapter.getSelectedItems().isEmpty()) {
            getMenuInflater().inflate(R.menu.activity_selected_opponents, menu);
        } else {
            getMenuInflater().inflate(R.menu.activity_opponents, menu);
        }

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        switch (id) {
            case R.id.update_opponents_list:
                updateDialogsAdapter();
                loadUsersFromQb();
                return true;

            case R.id.settings:
                showSettings();
                return true;

            case R.id.log_out:
                logOut();
                return true;

            case R.id.start_video_call:
                isVideoCall = true;
                startConference();
                return true;

            case R.id.start_audio_call:
                isVideoCall = false;
                startConference();
                return true;

            case R.id.start_as_listener:
                isVideoCall = true;
                startConference(dialogID, currentUser.getId(), isVideoCall, occupants, true);
                return true;
            case R.id.menu_appinfo:
                AppInfoActivity.start(this);
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void startConference() {
        if (checker.lacksPermissions(Consts.PERMISSIONS)) {
            startPermissionsActivity(!isVideoCall);
        } else {
            startConference(dialogID, currentUser.getId(), isVideoCall, occupants, false);
        }
    }

    private void updateDialogsAdapter() {
        startLoadDialogs();
    }

    private void showSettings() {
        SettingsActivity.start(this);
    }

    private void logOut() {
        UsersUtils.removeUserData(getApplicationContext());
        requestExecutor.signOut();
        startLoginActivity();
    }

    private void startLoginActivity() {
        LoginActivity.start(this);
        finish();
    }

    public void onCreateNewDialog(View view) {
        SelectUsersActivity.startForResult(this, REQUEST_SELECT_PEOPLE);
    }

    @SuppressWarnings("unchecked")
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK) {
            if (requestCode == REQUEST_SELECT_PEOPLE) {
                ArrayList<QBUser> selectedUsers = (ArrayList<QBUser>) data
                        .getSerializableExtra(SelectUsersActivity.EXTRA_QB_USERS);

                ProgressDialogFragment.show(getSupportFragmentManager(), R.string.create_dialog);
                createDialog(selectedUsers);
            }
            if (requestCode == REQUEST_PERMISSION) {
                startConference(dialogID, currentUser.getId(), isVideoCall, occupants, false);
            } else {
                updateDialogsAdapter();
            }
        }
    }

    private void createDialog(final ArrayList<QBUser> selectedUsers) {
        requestExecutor.createDialogWithSelectedUsers(selectedUsers, currentUser,
                new QBEntityCallback<QBChatDialog>() {
                    @Override
                    public void onSuccess(QBChatDialog dialog, Bundle args) {
                        //                        dialogsManager.sendSystemMessageAboutCreatingDialog(systemMessagesManager, dialog);
                        Log.d(TAG, "createDialogWithSelectedUsers dialog name= " + dialog.getName());
                        updateDialogsAdapter();
                        ProgressDialogFragment.hide(getSupportFragmentManager());
                    }

                    @Override
                    public void onError(QBResponseException e) {
                        ProgressDialogFragment.hide(getSupportFragmentManager());
                        showErrorSnackbar(R.string.dialogs_creation_error, null, null);
                    }
                }
        );
    }

    private class DeleteActionModeCallback implements ActionMode.Callback {

        public DeleteActionModeCallback() {
            fab.hide();
        }

        @Override
        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            mode.getMenuInflater().inflate(R.menu.activity_selected_dialogs, menu);
            return true;
        }

        @Override
        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            return false;
        }

        @Override
        public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
            switch (item.getItemId()) {
                case R.id.delete_dialog:
                    deleteSelectedDialogs();
                    if (currentActionMode != null) {
                        currentActionMode.finish();
                    }
                    return true;
            }
            return false;
        }

        @Override
        public void onDestroyActionMode(ActionMode mode) {
            currentActionMode = null;
            dialogsAdapter.clearSelection();
            invalidateOptionsMenu();
            fab.show();
        }


        private void deleteSelectedDialogs() {
            final Collection<QBChatDialog> selectedDialogs = dialogsAdapter.getSelectedItems();
            requestExecutor.deleteDialogs(selectedDialogs, new QBEntityCallback<ArrayList<String>>() {
                @Override
                public void onSuccess(ArrayList<String> dialogsIds, Bundle bundle) {
                    updateDialogsAdapter();
                }

                @Override
                public void onError(QBResponseException e) {
                    showErrorSnackbar(R.string.dialogs_deletion_error, e,
                            new View.OnClickListener() {
                                @Override
                                public void onClick(View v) {
                                    deleteSelectedDialogs();
                                }
                            });
                }
            });
        }
    }

    private void startConference(final String dialogID, int userID, boolean isVideoCall, final List<Integer> occupants, final boolean asListener) {
        Log.d(TAG, "startConference()");
        ProgressDialogFragment.show(getSupportFragmentManager(), R.string.join_conference);
        ConferenceClient client = ConferenceClient.getInstance(getApplicationContext());

        QBRTCTypes.QBConferenceType conferenceType = isVideoCall
                ? QBRTCTypes.QBConferenceType.QB_CONFERENCE_TYPE_VIDEO
                : QBRTCTypes.QBConferenceType.QB_CONFERENCE_TYPE_AUDIO;

        client.createSession(userID, conferenceType, new ConferenceEntityCallback<ConferenceSession>() {
            @Override
            public void onSuccess(ConferenceSession session) {
                ProgressDialogFragment.hide(getSupportFragmentManager());
                webRtcSessionManager.setCurrentSession(session);
                Log.d(TAG, "DialogActivity setCurrentSession onSuccess() session getCurrentUserID= " + session.getCurrentUserID());

                CallActivity.start(DialogsActivity.this, dialogID, occupants, asListener);
            }

            @Override
            public void onError(WsException responseException) {
                ProgressDialogFragment.hide(getSupportFragmentManager());
                if (!TextUtils.isEmpty(responseException.getMessage())) {
                    Exception exception = new Exception(responseException.getMessage());
                    showErrorSnackbar(R.string.join_conference_error, exception, null);
                } else {
                    showErrorSnackbar(R.string.join_conference_error, null, null);
                }
            }
        });
    }
}

File: sample-videochat-conference-kotlin/src/main/java/com/quickblox/sample/videochat/conference/kotlin/activities/DialogsActivity.kt
package com.quickblox.sample.videochat.conference.kotlin.activities

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.view.ActionMode
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.quickblox.chat.model.QBChatDialog
import com.quickblox.conference.ConferenceClient
import com.quickblox.conference.ConferenceSession
import com.quickblox.conference.WsException
import com.quickblox.conference.callbacks.ConferenceEntityCallback
import com.quickblox.core.QBEntityCallback
import com.quickblox.core.exception.QBResponseException
import com.quickblox.sample.videochat.conference.kotlin.R
import com.quickblox.sample.videochat.conference.kotlin.adapter.DialogsAdapter
import com.quickblox.sample.videochat.conference.kotlin.db.QbUsersDbManager
import com.quickblox.sample.videochat.conference.kotlin.util.*
import com.quickblox.sample.videochat.conference.kotlin.utils.PREF_CURRENT_ROOM_NAME
import com.quickblox.sample.videochat.conference.kotlin.utils.SharedPrefsHelper
import com.quickblox.sample.videochat.conference.kotlin.utils.WebRtcSessionManager
import com.quickblox.sample.videochat.conference.kotlin.utils.shortToast
import com.quickblox.users.model.QBUser
import com.quickblox.videochat.webrtc.QBRTCTypes

class DialogsActivity : BaseActivity() {
    private val PERMISSIONS = arrayOf(Manifest.permission.CAMERA, Manifest.permission.RECORD_AUDIO)

    private lateinit var dialogsRecyclerView: RecyclerView
    private lateinit var fab: FloatingActionButton
    private lateinit var dialogID: String
    private lateinit var occupants: MutableList<Int>

    private var currentActionMode: ActionMode? = null
    private var dialogsAdapter: DialogsAdapter? = null
    private var currentUser: QBUser? = null
    private var isVideoCall: Boolean = false


    companion object {
        private const val PERMISSION_CODE = 177
        private const val REQUEST_USERS_CODE = 174

        fun start(context: Context) {
            context.startActivity(Intent(context, DialogsActivity::class.java))
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_dialogs)

        currentUser = SharedPrefsHelper.getQbUser()

        initActionBar()
        initUI()
        startLoadDialogs()
        loadUsersFromQb()
    }

    private fun initActionBar() {
        val currentRoomName = SharedPrefsHelper[PREF_CURRENT_ROOM_NAME, ""]

        supportActionBar?.title = currentRoomName

        val currentUserFullName = currentUser?.fullName as String
        supportActionBar?.subtitle = String.format(getString(R.string.subtitle_text_logged_in_as), currentUserFullName)
    }

    override fun startSupportActionMode(callback: ActionMode.Callback): ActionMode? {
        currentActionMode = super.startSupportActionMode(callback) as ActionMode
        return currentActionMode
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val items = dialogsAdapter?.getSelectedItems()
        if (items != null && items.isNotEmpty()) {
            menuInflater.inflate(R.menu.activity_selected_opponents, menu)
        } else {
            menuInflater.inflate(R.menu.activity_opponents, menu)
        }
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        when (id) {
            R.id.update_opponents_list -> {
                startLoadDialogs()
                loadUsersFromQb()
                return true
            }
            R.id.settings -> {
                SettingsActivity.start(this)
                return true
            }
            R.id.log_out -> {
                removeAllUserData()
                return true
            }
            R.id.appinfo -> {
                AppInfoActivity.start(this)
                return true
            }
            R.id.start_video_call -> {
                isVideoCall = true
                startConference()
                return true
            }
            R.id.start_audio_call -> {
                isVideoCall = false
                startConference()
                return true
            }
            R.id.start_as_listener -> {
                isVideoCall = true
                startConference(dialogID, currentUser?.id as Int, isVideoCall, occupants, true)
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode != Activity.RESULT_OK) {
            return
        }
        if (requestCode == REQUEST_USERS_CODE) {
            val selectedUsers =
                    data?.getSerializableExtra(SelectUsersActivity.RESULT_USERS_KEY) as ArrayList<QBUser>
            showProgressDialog(R.string.dlg_create_dialog)
            createDialog(selectedUsers)
        }
        if (requestCode == PERMISSION_CODE) {
            val userId = currentUser?.id ?: return
            startConference(dialogID, userId, isVideoCall, occupants, false)
        }
    }

    private fun loadUsersFromQb() {
        showProgressDialog(R.string.dlg_loading_dialogs_users)
        val currentRoomName = SharedPrefsHelper.get<String>(PREF_CURRENT_ROOM_NAME)
        loadUsersByTag(currentRoomName, object : QBEntityCallback<ArrayList<QBUser>> {
            override fun onSuccess(result: ArrayList<QBUser>, params: Bundle) {
                hideProgressDialog()
                QbUsersDbManager.saveAllUsers(result, true)
            }

            override fun onError(responseException: QBResponseException) {
                hideProgressDialog()
                showErrorSnackbar(R.string.loading_users_error, responseException, View.OnClickListener { loadUsersFromQb() })
            }
        })
    }

    private fun removeAllUserData() {
        SharedPrefsHelper.clearAllData()
        QbUsersDbManager.clearDB()

        signOut()
        showLoginScreen()
    }

    private fun showLoginScreen() {
        LoginActivity.start(this)
        finish()
    }

    private fun initUI() {
        dialogsRecyclerView = findViewById(R.id.list_dialogs)
        fab = findViewById(R.id.fab_dialogs_new_chat)
        fab.setOnClickListener { SelectUsersActivity.startForResult(this, REQUEST_USERS_CODE) }
    }

    private fun startLoadDialogs() {
        showProgressDialog(R.string.dlg_loading_dialogs_users)
        loadDialogs(object : QBEntityCallback<ArrayList<QBChatDialog>> {
            override fun onSuccess(result: ArrayList<QBChatDialog>, params: Bundle) {
                hideProgressDialog()
                initDialogAdapter(result)
            }

            override fun onError(responseException: QBResponseException) {
                hideProgressDialog()
                showErrorSnackbar(R.string.loading_users_error, responseException, View.OnClickListener {
                    startLoadDialogs()
                })
            }
        })
    }

    private fun initDialogAdapter(chatDialogs: List<QBChatDialog>) {
        if (dialogsAdapter == null) {
            dialogsAdapter = DialogsAdapter(this, chatDialogs, object : DialogsAdapter.OnItemClickListener {
                override fun onShortClick(dialog: QBChatDialog) {
                    shortClickDialog(dialog)
                }

                override fun onLongClick(dialog: QBChatDialog) {
                    longClickDialog(dialog)
                }
            })
            dialogsRecyclerView.layoutManager = LinearLayoutManager(this)
            dialogsRecyclerView.adapter = dialogsAdapter
        } else {
            dialogsAdapter?.updateDialogs(chatDialogs)
        }
    }

    private fun shortClickDialog(dialog: QBChatDialog) {
        if (currentActionMode == null) {
            val occupants = dialog.occupants
            occupants.remove(currentUser?.id)
            dialogsAdapter?.toggleOneItem(dialog)
            invalidateOptionsMenu()
        } else {
            dialogsAdapter?.toggleSelection(dialog)
            updateActionBar(dialogsAdapter?.getSelectedItems()?.size)
        }
        occupants = dialog.occupants
        occupants.remove(currentUser?.id as Int)
        dialogID = dialog.dialogId
    }

    private fun longClickDialog(dialog: QBChatDialog) {
        fab.hide()
        startSupportActionMode(object : ActionMode.Callback {
            override fun onActionItemClicked(mode: ActionMode?, item: MenuItem?): Boolean {
                when (item?.itemId) {
                    R.id.delete_dialog -> {
                        deleteSelectedDialogs()
                        currentActionMode?.finish()
                        return true
                    }
                }
                return false
            }

            override fun onCreateActionMode(mode: ActionMode?, menu: Menu?): Boolean {
                mode?.menuInflater?.inflate(R.menu.activity_selected_dialogs, menu)
                return true
            }

            override fun onPrepareActionMode(mode: ActionMode?, menu: Menu?): Boolean {
                return false
            }

            override fun onDestroyActionMode(mode: ActionMode?) {
                currentActionMode = null
                dialogsAdapter?.clearSelectedItems()
                invalidateOptionsMenu()
                fab.show()
            }
        })
        dialogsAdapter?.selectItem(dialog)
        updateActionBar(dialogsAdapter?.getSelectedItems()?.size)
    }

    private fun updateActionBar(countSelectedUsers: Int?) {
        countSelectedUsers ?: return
        currentActionMode?.subtitle = null
        val title = if (countSelectedUsers > 1) {
            R.string.tile_many_dialogs_selected
        } else {
            R.string.title_one_dialog_selected
        }
        currentActionMode?.title = String.format(getString(title), countSelectedUsers)
        currentActionMode?.invalidate()
    }

    private fun deleteSelectedDialogs() {
        val selectedDialogs = dialogsAdapter?.getSelectedItems() ?: return
        deleteDialogs(selectedDialogs, object : QBEntityCallback<ArrayList<String>> {
            override fun onSuccess(dialogsIds: ArrayList<String>, bundle: Bundle) {
                startLoadDialogs()
            }

            override fun onError(e: QBResponseException) {
                showErrorSnackbar(R.string.dialogs_deletion_error, e,
                        View.OnClickListener { deleteSelectedDialogs() })
            }
        })
    }

    private fun createDialog(selectedUsers: ArrayList<QBUser>) {
        val currentUser = this.currentUser ?: return
        createDialogWithSelectedUsers(selectedUsers, currentUser, object : QBEntityCallback<QBChatDialog> {
            override fun onSuccess(dialog: QBChatDialog, args: Bundle) {
                startLoadDialogs()
                hideProgressDialog()
            }

            override fun onError(e: QBResponseException) {
                hideProgressDialog()
                showErrorSnackbar(R.string.dialogs_creation_error, e, View.OnClickListener { })
            }
        }
        )
    }

    private fun startConference() {
        if (checkPermissions(*PERMISSIONS)) {
            PermissionsActivity.startForResult(this, PERMISSION_CODE, !isVideoCall, PERMISSIONS)
        } else {
            val userId = currentUser?.id ?: return
            startConference(dialogID, userId, isVideoCall, occupants, false)
        }
    }

    private fun startConference(dialogID: String, userID: Int, isVideoCall: Boolean, occupants: List<Int>, asListener: Boolean) {
        if (!NetworkConnectionChecker(application).isConnectedNow()) {
            shortToast(R.string.no_internet_connection)
            return
        }
        showProgressDialog(R.string.join_conference)
        val client = ConferenceClient.getInstance(applicationContext)
        val conferenceType = if (isVideoCall) {
            QBRTCTypes.QBConferenceType.QB_CONFERENCE_TYPE_VIDEO
        } else {
            QBRTCTypes.QBConferenceType.QB_CONFERENCE_TYPE_AUDIO
        }
        client.createSession(userID, conferenceType, object : ConferenceEntityCallback<ConferenceSession> {
            override fun onSuccess(session: ConferenceSession) {
                hideProgressDialog()
                WebRtcSessionManager.setCurrentSession(session)
                CallActivity.start(this@DialogsActivity, dialogID, occupants, asListener)
            }

            override fun onError(responseException: WsException) {
                hideProgressDialog()
                showErrorSnackbar(R.string.join_conference_error, responseException, null)
            }
        })
    }
}